### ISSUE TITLE
Enhance join prefilter optimizer

### ISSUE THREAD

**Issue by kaikalur** [prestodb/presto#24223]
Currently we only support it when the other side is scan. We should enhance this feature to:

a) union all of scan
b) if the side has an aggregation on the join key we can still do it (and not do distinct)
c) drop any rows where any of the keys is null (especially important for multi-key joins)
d) make it cost based so that we prefilter the "correct" side which gives best results 
* also do it for outer joins - flipping the sides and left/right outer join if needed

CC: @feilong-liu

**Comment by Shravanpmk** [comment]
If I am not wrong it should in the file : `presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/JoinPrefilter.java`

**Comment by Shravanpmk** [comment]
@kaikalur  Can you please guide me through this issue 
Would be of much help thanks

**Comment by kaikalur** [comment]
> @kaikalur Can you please guide me through this issue Would be of much help thanks

Yes you found the code. Basically, right now we do it only if the other side is only a scan+filter+project. So extend it for other cases that I listed above.

**Comment by Shravanpmk** [comment]
Right @kaikalur I will take it up accordingly . Sorry for this but I might need some help initially.
